### PR TITLE
Improve history API

### PR DESCRIPTION
**Pull request by ClementNerma**
This PR reworks the history API, and brings several improvements:

* In the `History` trait, adding and replacing entries are now two distinct methods, reducing confusion
* `HistoryItem` values always have an ID, reducing confusion
* `History` values are responsible for generating IDs for new entries
* `FileBackedHistory` now longer has duplicate IDs in the history

`FileBackedHistory` as well as `SqliteBackedHistory` now generate IDs randomly, using `rand`'s `SmallRng`. This RNG is created from a system-provided source of entropy during the history creation, and is designed to generate numbers quickly.

This PR induces two main downsides:

First, creating a `FileBackedHistory` or `SqliteBackedHistory` may now cause a slight delay for gathering entropy. I tested on two machines and got less than 1s for 1 million RNG creation, but there may be edge cases where it takes longer).

Also, the `History` API contains breaking changes. In the current context, I don't consider this to be a large problem as no one seems to use this API anyway - up until two days ago, it was impossible to implement your own `History` due to `HistoryItemId` not being constructable from the outside.

Finally, a new dependency is pulled: `rand`. It only uses the `small_rng` and `getrandom` features and so it's pretty minimal, but it's still another dependency. It is used for non-clashing history item IDs generation.

Overall, this PR aims to make the API easier to write, read and use, while also reducing confusion in how the history works. It also removes what could be a bug in the future in `FileBackedHistory` which previously could have duplicate IDs for entries.

### CODE REVIEW

**Comment by ClementNerma** [comment]
Hmmm seems like some tests don't pass because the IDs are expected to be contiguous. This wasn't mentioned in the docs, I'll see what can I do about this.

**Comment by sholderbach** [comment]
The goal of this is still a bit unclear to me, you describe changes but not what is problematic with the current system of ids.

Having the IDs private before #677 allowed us to guarantee that those possibly implicit guarantees are upheld.

Maybe worth writing out what the current IDs all mean, what requirements we impose to implement session isolation/cwd aware hints etc. The constraint of the existing text file format only storing the command needs to be upheld for Nushell backwards compatibility unless we have a clear story,

**Comment by ClementNerma** [comment]
Before I go further and rework the histories' internals, @fdncred may I have your opinion on this?

File: src/engine.rs
Comment (ClementNerma) [reply src/engine.rs:1759]: I just left how it was before the change :p
But in the future we should do a proper error handling, yes.

**Comment by ClementNerma** [comment]
@fdncred and @sholderbach thanks for your answers!

> The goal of this is still a bit unclear to me, you describe changes but not what is problematic with the current system of ids.
> 
> Having the IDs private before #677 allowed us to guarantee that those possibly implicit guarantees are upheld.
> 
> Maybe worth writing out what the current IDs all mean, what requirements we impose to implement session isolation/cwd aware hints etc. The constraint of the existing text file format only storing the command needs to be upheld for Nushell backwards compatibility unless we have a clear story,

Having the IDs private prevented anyone to make a custom `History`. Which is a problem, because having a basic file or SQLite implementation doesn't cover anyone's needs.

For me, using random IDs is the best choice: we don't make _any_ assumption on what an ID represents - it is only that, an ID. Not an index, or anything else.

It can be generated in a variety of manners depending on the needs. For an in-memory history for instance, it could be a simple counter. In most cases, a random generator (like I used) would be the best choice to avoid ID re-use.

With this PR we have a system where IDs are black boxes, and are only used to refer to a specific history item - nothing more.

Navigation in the history, ordering etc. is the responsibility of the `History` itself.

This will allow anyone to build a custom `History` without having to worry about:

1. How the IDs are generated
2. How they can handle
3. What are their guarantees
4. What guarantees the navigation system (`History::navigate`) must satisfy based on the provided IDs

**Comment by sholderbach** [comment]
Mhh to me what the capabilities and requirements for your `History` implementation are, is still very vague.
What would prevent you from implementing that inside `reedline` itself, where we can more easily control the assumptions of the overall interface, before preeemptively making everything public, making it much harder to reason about what has to hold for all the other components to properly work.

Reedline needs to improve in this area for Nushell as well, but designing against a vague outside target requires us to dedicate time to reason about much more as to not box in our Nushell-internal design space. From our perspective having a locked trait is perfectly fine.
It seems based on your recent PRs that you don't yet have a working `History` implementation due to the private things, maybe worth spinning this up in your reedline fork for us to take a shared look at after it satisfies your requirements.

**Comment by ClementNerma** [comment]
> Mhh to me what the capabilities and requirements for your `History` implementation are, is still very vague. What would prevent you from implementing that inside `reedline` itself, where we can more easily control the assumptions of the overall interface, before preeemptively making everything public, making it much harder to reason about what has to hold for all the other components to properly work.

Because why keep anything private when we can make it customizable from the outside? The PR I'm proposing doesn't make the history system anymore complex than it already is ; in fact I'd argue it is simpler this way. IDs are black boxes and we don't make any assumption about that - that's simpler than making assumptions everywhere about what they mean.
 
> Reedline needs to improve in this area for Nushell as well, but designing against a vague outside target requires us to dedicate time to reason about much more as to not box in our Nushell-internal design space. From our perspective having a locked trait is perfectly fine. It seems based on your recent PRs that you don't yet have a working `History` implementation due to the private things, maybe worth spinning this up in your reedline fork for us to take a shared look at after it satisfies your requirements.

This PR will contain the updated code for `FIleBackedHistory` notably (I'm currently working on it), resolving multiple issues this struct had, notably IDs clashing. I had several situations where the history didn't keep track of some informations because of that.

**Comment by ClementNerma** [comment]
CI now passes 🥳 

So, there is one big problem with this PR: the underlying data (text file for `FileBackedHistory` and database file for `SqliteBackedHistory`) have changed. Text file has a slightly different format (to store IDs), and the database has one new column to store the indexes, required to order rows.

This means there needs to be a migration for older files - this can be done easily with a simple function call though. But this needs a migration is required. I'm currently searching for a solution to avoid this.

**Comment by ClementNerma** [comment]
Ok so I managed to resolve the problem for the SQLite database: by default the `user_version` pragma is always `0` and it is always retrieved during startup.

So I perform a simple check:
* If `user_version = 0` (the default):
  * If the `history` table does not exist, we create it using the new schema
  * Otherwise we migrate its data to the new schema, keeping existing IDs
  * Then we set `user_version` to `1`
* If `user_version = 1`, this means the schema exists and is the latest version
* If it's another value, we fail (just like before)

This way, existing databases will be migrated seamlessly and there won't be the slightest performance hit once the migration has been done.

**Comment by ClementNerma** [comment]
The PR is now complete. Some tests still need to be added to check edge cases and how the migration behaves in different scenarios, but all the changes and features are fully implemented.

To sum up a little bit: the main change is the `History` trait changed. This led to changes in how `FileBackedHistory` and `SqliteBackedHistory` behave, which then led to a change in their underlying format (text file format / SQL schema).

Concretely, the new system should be invisible to users, apart from the API change (which is trivial to adapt if they want to keep the same behaviour as they currently have).

Both history structs now use an internal [`rand::rngs::SmallRng`](https://docs.rs/rand/latest/rand/rngs/struct.SmallRng.html) whose entropy is gathered from the operating system. This allows to generate random numbers quickly, more than with a cryptographically-secure RNG which is not required in the current context. Given the `History` trait is responsible for the generation, anyone can use its own RNG if required, but the default should cover almost all possible cases.

A slight initialization cost will be induced by the entropy gathering, but it should be quick enough to be unnoticeable (less than 10 µs (microseconds) average on two of my machines in release mode).

Performance will be slightly lower given the additional need to handle IDs, but it should be low enough to be entirely unnoticeable. Some benchmarks could be conducted on these changes but I'm confident it won't be perceptible to end users.

Safety is increased ; some long-standing bugs have been resolved in this PR, including history truncating creating IDs collision. String escaping has also been added to the SQLite backend when searching with a pattern (using a `LIKE` pattern).

History backends still mostly behave as they used to: `FileBackedHistory` still uses a line-based text file format, and `SqliteBackedHistory` still uses a SQLite database with a single `history` table and the same columns, aside from one new column used for indexing.

`FileBackedHistory` gets a compatibility layer ; it now uses a new line format to include the ID inside, but will still support the old file formats without any changes. Old-format and new-format lines can be mixed together inside the same file.

`SqliteBackedHistory` gets a migration process ; the first time the database is opened, it will migrate the previous history data to a new one using a transaction. This will only happen once, and should be almost instantaneous even with a large history. Interruption of the migration process should not leave the history in a corrupted state, as the transaction will rollback.

Tests are mostly unchanged, they have been slightly updated to reflect the new API but are otherwise identical - so the code coverage doesn't changes and the tests are still as reliable as previously.

An interesting add in the future could be to add, either as a builtin or as an example, an `History` backed by a [`sled`](https://docs.rs/sled/latest/sled/) database - which is perfectly fit for this case, more than both a text file or an SQLite database. The existing structs will be kept anyway for compatibility, especially considering the `SqliteBackedHistory` is currently used by `nushell`.

@fdncred and @sholderbach may I have your opinion on this? If you agree with it, I'll write the new tests to ensure everything is working perfectly (notably the migration process), and then we'll be able to merge this PR.

**Review by fdncred — CHANGES_REQUESTED** [review]
Thanks for taking the time to try and improve reedline and for the summary at the bottom, it helps.

I see some good work in here that on some level makes dealing with history more intuitive. I also see some things that I don't agree with and/or have questions about. 

I'm really concerned about landing this on many levels. Here's some of what I'd need to feel better about in order for me to vote yes to land this PR. I'm not quite sure how to get to a good level of comfort. Suggestions welcome.
- I need to feel good about how it's _NOT_ going to affect nushell users.
- I need to feel certain that the changes made here are going in the right direction of where we want to take reedline in the future.
- I need to feel like this is better than what it was before.
- I need to feel like our tests cover as much as possible in order to allay my concerns.
- I think we need to have some consensus of a few nushell core-team members (up/down vote) since this is such a large PR and really changes everything related to history.

**Comment by ClementNerma** [comment]
> I need to feel good about how it's NOT going to affect nushell users.

This PR is designed to not affect everyone in a perceptible manner. All the changes are hunder the hood. The only difference is the API revamp which needs a little change (concretely, we're talking propably less than a minute of code update, that's really a detail).

> I need to feel certain that the changes made here are going in the right direction of where we want to take reedline in the future.

That entirely depends on your vision of the project, but I feel like reedline would be the best at being both a reference implementation and solid ground for Nushell **AND** being a nice line editing library to use in custom shells.

> I need to feel like this is better than what it was before.

I feel it is way better, but that's a personal take. We now have a clearer ID system, less error-prone, a cleaner API.

> I need to feel like our tests cover as much as possible in order to allay my concerns.

Absolutely, which is why I said I would write the new tests if this PR is approved (what I mean by that is if you're okay with merging the PR **at the condition** of it being correctly tested, I'll write the tests).

> I think we need to have some consensus of a few nushell core-team members (up/down vote) since this is such a large PR and really changes everything related to history.

Well, yes and no actually. It mostly works like previously, it's mostly under-the-hood changes. Especially the SQLite backend, which Nushell uses, uses an almost identical system: you just have one more column for indexes, and that's all.

But I understand your concern and given reedline's under the "reponsability" of Nushell it would be a good thing to get an approval from Nushell's team.

File: src/history/sqlite_backed.rs
Comment (ClementNerma) [reply src/history/sqlite_backed.rs:102]: They **should** use it, but this hasn't been implemented yet. For now I've left these parts untouched.
I feel like changing their behaviour would be a bad idea right now as it would change how Nushell behaves.

The method separation is useful for clarifying behaviour, and here we can clearly see a "feature" is missing, while it wasn't obvious at all previously.

For histories that support replacing, this will be useful. I added it because reedline requires histories to replace history items. It could have been a call to `update`, but it's how reedline is designed, so I preferred to not change that.

File: src/history/sqlite_backed.rs
Comment (ClementNerma) [reply src/history/sqlite_backed.rs]: Seems like a typo indeed, inherited from the current codebase.
Fixed

File: src/history/sqlite_backed.rs
Comment (ClementNerma) [reply src/history/sqlite_backed.rs]: Yes indeed, it was a typo from the codebase, just fixed it.

File: src/history/file_backed.rs
Comment (fdncred) [reply src/history/file_backed.rs:99]: I'm not objecting to getting the first ID somewhere. I'm objecting to using unwrap. I'd guess that you can match on it and handle errors and use map_err if the error is a different flavor.

Comment (ClementNerma) [reply src/history/file_backed.rs:99]: This would be misleading. Using a `.map_err()` in a situation we've _proven_ to meet the condition required for unwrapping may lead other developers to think there is a case where this could fail - while it can't, that's why I used .unwrap() (which I always use very sparingly).

Writing `.first().unwrap()` is more readable, less confusing, and more correct as if we don't meet this condition something's **_very_** wrong and we cannot continue processing anything because some of the code is completely invalid.

File: src/history/sqlite_backed.rs
Comment (ClementNerma) [reply src/history/sqlite_backed.rs:284]: We absolutely use the `id` column: that's the one used for inserting, updating, removing, and fetching indexes. Everything is based on that column, `idx` is only used for a specific situation: when we want to query >1 items from an existing one in either direction.

Comment (fdncred) [reply src/history/sqlite_backed.rs:284]: ok, good. maybe it's the name that i'm hung up on then. id and idx are very similar. maybe `query_idx` or something similar. it's just not intuitive having such similar names.

**Comment by ClementNerma** [comment]
> > Well, yes and no actually.
> 
> Clearly, this is a breaking change and changes a ton of guts. The history is all dealt with differently now that you've added a new index column. You try to maintain backwards compatibility as much as possible, which is appreciated, but any reedline user will have to deal with your changes if we accept this PR. There's no getting around it. I understand your point though.

This PR is designed to make changes seamless. Everything that worked previously still works now. If we list all the things that change for users, it is:

* The new (breaking) API for `History`
* The fact that the new SQLite schema and the new text file format can contain new informations
* The fact that the SQLite database now uses unordered IDs

No one has implemented a custom `History` before given the previous visibility of some items, so the only real problem that may arise is the fact the text file format changes (the SQLite schema is backwards-compatible after all, aside from the ID column).

But I _think_ it is such a small number of users with such a minimal impact that this isn't a major problem ; it would be a bigger problem if Nushell used the text file format as some tools may have relied on it ; but fortunately it uses the SQLite database instead.

**Comment by stormasm** [comment]
@ClementNerma we are happy you are taking the time to review and understand our code better...

It is going to take our team some time to digest and understand the ramifications of your changes as well as have some internal discussions about the work you have done so that we can move forward and give you some feedback.

I will review your code as well and get a better understanding of it...

My first question for you is this ?

In our current history implementation by default we save the same entry numerous times.

For example if I type

```rust
version
ls
version
ls
version
ls
history
ls
history
ls
```

We will save all of those *ls* entries as well as all of the *history* entries and *version* entries.

I would like to have one *ls* and one *history* and one *version* across my entire history file...

How easy would it be to do this with your current solution just out of curiosity ?

Thanks for your work --- we need folks like yourself who are willing to take on the work and stay with it over time...

Our concern is that you come in and make a bunch of changes which we are not totally able to understand the motivation for them and then we are responsible over the longer term maintaining that code...

Hope you understand and are willing to work with us on the *reedline* communities behalf.

**Comment by stormasm** [comment]
@ClementNerma I reviewed your PR and now have a better understanding of the work you have done...

First of all let me say that @sholderbach is the lead person on reedline and @fdncred also knows a lot about reedline but will defer to @sholderbach for final approval on anything major like this PR.

I am just another member of the nushell core team but with no expertise on reedline...

As I stated in my previous comments we are happy to have folks like yourself who are willing to contribute to the reedline community but you should understand that we are incredibly stretched in our ability to manage all of the work that we have to deal with both in the reedline community and nushell community as a whole.

With that said the chance of this PR landing is low.  

I am not going to close it because that would be up to @sholderbach or @fdncred to make the final call but in order for you to be able to be a valuable member of the community we are going to have to take a different approach.

We want you to help us maintain and make reedline better over the long term so don't take this the wrong way...

Our team is much happier with smaller more focused PRs with an end design goal that has been well thought out and discussed prior...

So I would suggest entering into more discussion in the reedline discord channel with your ideas and possibly maintaining reedline by assisting with and commenting on PRs as they come in...

Overall I like your idea of having swappable History Implementations on top of a well defined History API but in order to get there its going to take a crawl, walk, run approach which is how we like to phrase it...

Also I personally do not like having the HistoryItem IDs locked down and fixed --- The History Items IDs should be created dynamically  each time Reedline fires up...

And storing the ItemID's in the history.txt file is just not good IMHO.

It bloats the size of the *history.txt* file dramatically and makes it almost impossible to post process the history.txt file if someone wants to use it for other purposes.  

Having the very clean, nice, simple *history.txt* file with just the command is the way to go...

**Comment by ClementNerma** [comment]
> As I stated in my previous comments we are happy to have folks like yourself who are willing to contribute to the reedline community but you should understand that we are incredibly stretched in our ability to manage all of the work that we have to deal with both in the reedline community and nushell community as a whole.
> ...
> Our team is much happier with smaller more focused PRs with an end design goal that has been well thought out and discussed prior...

I understand that, after all its just a FOSS project, not a paid product.

> So I would suggest entering into more discussion in the reedline discord channel with your ideas and possibly maintaining reedline by assisting with and commenting on PRs as they come in...
> 
> Overall I like your idea of having swappable History Implementations on top of a well defined History API but in order to get there its going to take a crawl, walk, run approach which is how we like to phrase it...

Seems like a good idea!

> Also I personally do not like having the HistoryItem IDs locked down and fixed --- The History Items IDs should be created dynamically each time Reedline fires up...

Allow to me disagree on that. IDs are used to differentiate items, how can you re-generate them on-the-fly? Don't forget entries can be inserted by another concurrent process interacting with the same history file / database.

> And storing the ItemID's in the history.txt file is just not good IMHO.
> 
> It bloats the size of the _history.txt_ file dramatically and makes it almost impossible to post process the history.txt file if someone wants to use it for other purposes.
> 
> Having the very clean, nice, simple _history.txt_ file with just the command is the way to go...

I totally agree with that, but it's hard to keep it simple. How do you identify uniquely a set of entries created during runtime, knowing that you may later have to synchronize the file to disk, discover there are new entries, and insert them at a previous position in the list? We need to have something that is entirely unique to each entry, which is not possible with a plain file.

If for instance you have one `ls` in memory, synchronize to disk, and find another `ls` in a previous position - how do you identify each one?

**Comment by stormasm** [comment]
@ClementNerma I am going to give you some preliminary ideas here --- this is just a starting point so please don't start coding yet --- lets just talk about a design going forward...

You might want to open up an issue and we can talk more about this there...

Here are my preliminary thoughts on this...

History - Menus - Completer can be referred to as *HMC* components

The Big Idea concept is a Complete Refactor / Rewrite of Reedline without having History / Menus / Completer (HMC)
so tightly embedded into the Engine ---- but instead API entry points into the Engine which would enable
decoupled HMC components...

>> What if HistoryItemId were only identifying for a single session? This would allow to greatly simplify data storage.

Yes --- I like this idea / most importantly greatly simplify the storage...

I want the whole history API to be a drop in replacement to Reedline in the future --- that is the end futuristic goal...

So imagine developers coming along and having the ability to write their own History implementation and it would
attach to our simple History API which is versatile...

Now moving on my "grand vision" for Reedline is to have the ability to have a simple Reedline with
* No History
* No Menus
* No Completers

Based on an API system that allowed you to Shut OFF History / Menus / Completers

This is for example purposes only --- most people would not want to shut it off but it shows
how you can configure Reedline with your own components or a lightweight Reedline with no components...

And or drop in your own History / Menus / Completers based on a well thought out API that enabled gluing on to a core Engine that accepted these components...

That is where I would like to end up...

The History API could be step one --- but it would entail lots of design to be able to pull it out of the Engine
and put back in History API entry points...

You could create another Reedline-History crate in your own repo to do the experimentation or something to that effect...

Please start an issue so we can continue the discussion there....

Not sure if you are up for this project --- but I think it could be a cool longer term effort that would take a while to implement...

Open to all sorts of ideas / strategies / criticisms of this concept --- I am just throwing ideas out there to see what people think...

Nu-cli should eventually be able to accept the alternative Reedline as it is being developed and then eventually
swap out the new Reedline with the current Reedline version...

Just so you can see my ideas....
@maxomatic458 @ysthakur @fdncred

**Comment by stormasm** [comment]
Just writing stuff down here as an example...

If you look at the [basic example](https://github.com/nushell/reedline/blob/main/examples/basic.rs) in Reedline...

```rust
let sig = line_editor.read_line(&prompt)?;
```

Why does the read_line method have to take a prompt ---- I never understood that design choice ?

**Comment by ysthakur** [comment]
@stormasm That would require the prompt to be stored inside the Reedline object, so mutating it from outside would become a bit tougher. It would definitely be convenient if you don't want to mutate the prompt, though.

**Comment by stormasm** [comment]
@ysthakur thank you !  for thinking about the prompt topic --- and updating the idea 😄

**Comment by Cactus-man** [comment]
@ysthakur GIven that `Prompt` is a _trait_ and not a structure with strings, it feels odd to change the actual prompt object. The trait methods are called every time it's rendered anyway. If you look at `examples/custom_prompt.rs`. You could mutate internal prompt state and branch the current prompt based on that (or even include such information in the prompt e.g. current time).

Yes, I think the current prompt API is not working to make that easy (you need to use interior mutability, unfortunately), but it probably is "the right thing to do".

**Pull request closed without merging**
